### PR TITLE
Release react 5.5.0

### PR DESCRIPTION
**Pull request by aaronlademann-wf**
This release of react includes...

__New Features__

- 🎉 🎉 🎉 __Support for function components, memo and hooks!!!__ 🎉 🎉 🎉

    Sooooo much work from so many amazing people made this possible, but to summarize: 
    
    - #221 Add support for function components
    - #252 Add support for `memo` higher order component
    - Hooks, hooks, and more hooks!
        - #223 useState
        - #233 useCallback
        - #237 useContext
        - #227 useEffect
        - #248 useLayoutEffect
        - #232 useReducer
        - #245 useRef
        - #250 useMemo
        - #247 useImperativeHandle
        - #246 useDebugValue

    <p><br>It works like this...</p>
    
    Define the component:
    ```dart
    import 'package:react/react.dart' as react;
    
    final SomeWidget = react.registerFunctionComponent(_SomeWidget, displayName: 'SomeWidget');
    
    _SomeWidget(Map props) {
      // You can use hooks in here, too!
      
      return react.div({}, [
        // Some children...
      ]);
    }
    ```
    
    Render the component _(exact same consumer API as a class-based component)_:
    ```dart
    import 'package:react/react_dom.dart' as react_dom;
    import 'some_widget.dart'; // Where your component is defined
    
    main() {
      final renderedWidget = SomeWidget({
        // put some props here
      }, [
        // put some children here!
      ]);
    
      react_dom.render(renderedWidget, querySelector('#idOfSomeNodeInTheDom'));
    }
    ```
    
    > Check out all the [function component and hooks examples](https://github.com/cleandart/react-dart/blob/c9a1211d5d77a9e354b864e99ef8f52b60eeff85/example/test/function_component_test.dart) for more information!
        
__Fixes / Updates__
- #253 Deprecate `setClientConfiguration`.
    - It is no longer necessary - and can be removed from your implementations
- #273 Make `JsBackedMap.values` compatible with MSIE 11

---

Changesets committed directly to the WIP branch that need review:
1. https://github.com/cleandart/react-dart/pull/265/files/cf43b36..066c138
2. https://github.com/cleandart/react-dart/pull/265/files/9714657..75d6611

(Click on "Changes from X commits..." to see commit messages for more context)
Screen Shot 2020-07-29 at 12 19 04 PM

### CODE REVIEW

File: lib/react_client/private_utils.dart [hunk 1]
```diff
@@ -0,0 +1,12 @@
+@JS()
+library react_client.private_utils;
```
Comment (greglittlefield-wf) [inline lib/react_client/private_utils.dart:2]: I think this should be under `src`; we don't seem to use it in over_react or anything

**Comment by semveraudit-wf** [comment]
### Public API Changes

 Recommendation: **:bangbang: Major version bump (fyi @Workiva/semver-audit-group )**


```diff
@@ line 6: package:react/src/prop_validator.dart @@
-  typedef PropValidator<T> = Error Function(T props, PropValidatorInfo info)
+  typedef PropValidator<TProps> = Error Function(TProps props, PropValidatorInfo info)
// `type` of `props` has changed.
// Changing a parameter signature is a major change.
```

```diff
@@ line 104: package:react/src/context.dart @@
-  Context createContext<TValue>([TValue defaultValue, int Function(TValue, TValue) calculateChangedBits])
+  Context<TValue> createContext<TValue>([TValue defaultValue, int Function(TValue, TValue) calculateChangedBits])
// `return_type` of `createContext` has changed from `Context` to `Context<TValue>`.
// Changing a top-level function is a major change.
```

```diff
@@ line 503: package:react/react.dart @@
abstract class Component2 implements Component
-     Context get contextType
+     Context<dynamic> get contextType
//    `type` of `contextType` has changed from `Context` to `Context<dynamic>`.
//    Changing a class field is a major change.

```

```diff
@@ line 839: package:react/react_client.dart @@
class ReactDomComponentFactoryProxy extends ReactComponentFactoryProxy
-     ReactDomComponentFactoryProxy ReactDomComponentFactoryProxy(dynamic name)
+     ReactDomComponentFactoryProxy ReactDomComponentFactoryProxy(String name)
//    `type` of `name` has changed.
//    Changing a parameter signature is a major change.

```

```diff
@@ line 125: package:react/react_client/react_interop.dart @@
-  ReactJsComponentFactoryProxy forwardRef(dynamic Function(Map<dynamic, dynamic>, Ref<dynamic>) wrapperFunction)
+  ReactJsComponentFactoryProxy forwardRef(dynamic Function(Map<dynamic, dynamic>, Ref<dynamic>) wrapperFunction, {String displayName = 'Anonymous'})
// `displayName` was added.
// Adding an optional parameter is a minor change.
```
<details>
<summary>Click to see 30 more API Changes</summary><p>

---
```diff
@@ line 2742: package:react/react.dart @@
-  void setReactConfiguration(dynamic domCreator, dynamic customRegisterComponent, {ReactDartComponentFactoryProxy2<Component2> Function(Component2 Function(), {Component2Bridge Function(Component2) bridgeFactory, Iterable<String> skipMethods}) customRegisterComponent2})
+  void setReactConfiguration(dynamic domCreator, dynamic customRegisterComponent, {ReactDartComponentFactoryProxy2<Component2> Function(Component2 Function(), {Component2Bridge Function(Component2) bridgeFactory, Iterable<String> skipMethods}) customRegisterComponent2, ReactDartFunctionComponentFactoryProxy Function(dynamic Function(JsBackedMap), {String displayName}) customRegisterFunctionComponent})
// `customRegisterFunctionComponent` was added.
// Adding an optional parameter is a minor change.
```

```diff
@@ line 125: package:react/react_client/react_interop.dart @@
-  ReactJsComponentFactoryProxy forwardRef(dynamic Function(Map<dynamic, dynamic>, Ref<dynamic>) wrapperFunction)
+  ReactJsComponentFactoryProxy forwardRef(dynamic Function(Map<dynamic, dynamic>, Ref<dynamic>) wrapperFunction, {String displayName = 'Anonymous'})
// `displayName` was added.
// Adding an optional parameter is a minor change.
```

```diff
@@ line 1968: package:react/react.dart @@
+  ReactDartFunctionComponentFactoryProxy Function(dynamic Function(JsBackedMap), {String displayName}) registerFunctionComponent
// Adding a top-level variable is a minor change.
```

```diff
@@ line 670: package:react/react_client/react_interop.dart @@
+  dynamic get jsNull
// Adding a top-level variable is a minor change.
```

```diff
@@ line 395: package:react/hooks.dart @@
+  Ref<T> useRef<T>([T initialValue])
// Adding a top-level function is a minor change.
```

```diff
@@ line 293: package:react/react_client/react_interop.dart @@
+  ReactJsComponentFactoryProxy memo(ReactDartFunctionComponentFactoryProxy factory, {bool Function(Map<dynamic, dynamic>, Map<dynamic, dynamic>) areEqual})
// Adding a top-level function is a minor change.
```

```diff
@@ line 293: package:react/react_client/react_interop.dart @@
+  ReactJsComponentFactoryProxy memo(ReactDartFunctionComponentFactoryProxy factory, {bool Function(Map<dynamic, dynamic>, Map<dynamic, dynamic>) areEqual})
// Adding a top-level function is a minor change.
```

```diff
@@ line 143: package:react/hooks.dart @@
+  void useEffect(dynamic Function() sideEffect, [List<Object> dependencies])
// Adding a top-level function is a minor change.
```

```diff
@@ line 330: package:react/hooks.dart @@
+  T useCallback<T extends Function>(T callback, List<dynamic> dependencies)
// Adding a top-level function is a minor change.
```

```diff
@@ line 428: package:react/hooks.dart @@
+  T useMemo<T>(T Function() createFunction, [List<dynamic> dependencies])
// Adding a top-level function is a minor change.
```

```diff
@@ line 33: package:react/react.dart @@
+  typedef DartFunctionComponent = dynamic Function(JsBackedMap props)
// Adding a typedef is a minor change.
```

```diff
@@ line 80: package:react/hooks.dart @@
+  StateHook<T> useState<T>(T initialValue)
// Adding a top-level function is a minor change.
```

```diff
+ react/react_client/zone.dart

// Adding an entry point is a minor change.
```

```diff
@@ line 62: package:react/react_client/component_factory.dart @@
+  Map<dynamic, dynamic> unconvertJsProps(dynamic instance)
// Adding a top-level function is a minor change.
```

```diff
+ react/react_client/component_factory.dart

// Adding an entry point is a minor change.
```

```diff
@@ line 47: package:react/react_client/component_factory.dart @@
+  Function unconvertJsEventHandler(Function jsConvertedEventHandler)
// Adding a top-level function is a minor change.
```

```diff
+ react/hooks.dart

// Adding an entry point is a minor change.
```

```diff
@@ line 240: package:react/hooks.dart @@
+  ReducerHook<TState, TAction, TInit> useReducer<TState, TAction, TInit>(TState Function(TState, TAction) reducer, TState initialState)
// Adding a top-level function is a minor change.
```

```diff
@@ line 675: package:react/react_client/react_interop.dart @@
+  dynamic get jsUndefined
// Adding a top-level variable is a minor change.
```

```diff
@@ line 532: package:react/hooks.dart @@
+  void useImperativeHandle(Ref<dynamic> ref, dynamic Function() createHandle, [List<dynamic> dependencies])
// Adding a top-level function is a minor change.
```

```diff
@@ line 294: package:react/hooks.dart @@
+  ReducerHook<TState, TAction, TInit> useReducerLazy<TState, TAction, TInit>(TState Function(TState, TAction) reducer, TInit initialArg, TState Function(TInit) init)
// Adding a top-level function is a minor change.
```

```diff
@@ line 16: package:react/src/typedefs.dart @@
+  typedef JsFunctionComponent = dynamic Function(JsMap props, [JsMap legacyContext])
// Adding a typedef is a minor change.
```

```diff
@@ line 27: package:react/react_client/component_factory.dart @@
+  dynamic listifyChildren(dynamic children)
// Adding a top-level function is a minor change.
```

```diff
@@ line 46: package:react/react.dart @@
+  typedef FunctionComponentRegistrar = ReactDartFunctionComponentFactoryProxy Function(dynamic Function(JsBackedMap) componentFactory, {String displayName})
// Adding a typedef is a minor change.
```

```diff
@@ line 95: package:react/react_client/js_interop_helpers.dart @@
+  Expando<bool> get isRawJsFunctionFromProps
// Adding a top-level variable is a minor change.
```

```diff
@@ line 606: package:react/hooks.dart @@
+  dynamic useDebugValue<T>(T value, [dynamic Function(T) format])
// Adding a top-level function is a minor change.
```

```diff
@@ line 103: package:react/hooks.dart @@
+  StateHook<T> useStateLazy<T>(T Function() init)
// Adding a top-level function is a minor change.
```

```diff
@@ line 359: package:react/hooks.dart @@
+  T useContext<T>(Context<T> context)
// Adding a top-level function is a minor change.
```

```diff
@@ line 464: package:react/hooks.dart @@
+  void useLayoutEffect(dynamic Function() sideEffect, [List<Object> dependencies])
// Adding a top-level function is a minor change.
```

```diff
@@ line 90: package:react/react_client/component_factory.dart @@
+  mixin JsBackedMapComponentFactoryMixin on ReactComponentFactoryProxy
// Adding a class is a minor change.

+     static JsMap generateExtendedJsProps(Map<dynamic, dynamic> props)
//    Adding a method is a minor change.

+     ReactElement build(Map<dynamic, dynamic> props, [List<dynamic> childrenArgs = const []])
//    Adding a method is a minor change.

      From package:react/react.dart
+     dynamic call(Map<dynamic, dynamic> props, [dynamic c1 = _notSpecified, dynamic c2 = _notSpecified, dynamic c3 = _notSpecified, dynamic c4 = _notSpecified, dynamic c5 = _notSpecified, dynamic c6 = _notSpecified, dynamic c7 = _notSpecified, dynamic c8 = _notSpecified, dynamic c9 = _notSpecified, dynamic c10 = _notSpecified, dynamic c11 = _notSpecified, dynamic c12 = _notSpecified, dynamic c13 = _notSpecified, dynamic c14 = _notSpecified, dynamic c15 = _notSpecified, dynamic c16 = _notSpecified, dynamic c17 = _notSpecified, dynamic c18 = _notSpecified, dynamic c19 = _notSpecified, dynamic c20 = _notSpecified, dynamic c21 = _notSpecified, dynamic c22 = _notSpecified, dynamic c23 = _notSpecified, dynamic c24 = _notSpecified, dynamic c25 = _notSpecified, dynamic c26 = _notSpecified, dynamic c27 = _notSpecified, dynamic c28 = _notSpecified, dynamic c29 = _notSpecified, dynamic c30 = _notSpecified, dynamic c31 = _notSpecified, dynamic c32 = _notSpecified, dynamic c33 = _notSpecified, dynamic c34 = _notSpecified, dynamic c35 = _notSpecified, dynamic c36 = _notSpecified, dynamic c37 = _notSpecified, dynamic c38 = _notSpecified, dynamic c39 = _notSpecified, dynamic c40 = _notSpecified])
//    Adding a method is a minor change.
+     dynamic get type
//    Adding a field is a minor change.

```
---

</p></details>

15 more changes could not be displayed on this comment. [Click here](https://semver-audit.wk-dev.wdesk.org/diffs/cleandart/react-dart/ae5a851ffc2080830180db8e407edd6aceee1f82/.../066c13863c6b54809a2947696e8746091c26251a) to view all changes.

Showing results for 066c13863c6b54809a2947696e8746091c26251a
> Powered by [semver-audit-service](https://github.com/Workiva/semver-audit-service). Please report any problems by [filing an issue](https://github.com/Workiva/semver-audit-service/issues).
> Reported by the dart semver audit client 2.2.0
[Browse public API](https://semver-audit.wk-dev.wdesk.org/cleandart/react-dart/066c13863c6b54809a2947696e8746091c26251a).

**Comment by greglittlefield-wf** [comment]
Semver changes look good to me, and should not actually be considered major. Specifically

- Passing any other type into `ReactDomComponentFactoryProxy ` would have resulted in a runtime error,
- Previously returning `Context<dynamic>` from `createContext` is considered a bug, and changing it shouldn't cause issues

File: lib/react.dart [hunk 2]
```diff
@@ -1296,23 +1296,6 @@ mixin TypedSnapshot<TSnapshot> {
...
-  /// ```
-  /// YourFactoryProxy.reactComponentFactory(props, children);
-  /// ```
-  ///
-  /// After:
-  /// ```
-  /// React.createElement(YourFactoryProxy.type, props, children);
-  /// ```
-  @Deprecated('6.0.0')
-  ReactJsComponentFactory reactComponentFactory;
```
Comment (greglittlefield-wf) [inline lib/react.dart]: This was added in this PR but was never used

File: lib/src/js_interop_util.dart [hunk 3]
```diff
@@ -1,10 +1,13 @@
 @JS()
-library react_client.private_utils;
```
Comment (greglittlefield-wf) [inline lib/src/js_interop_util.dart]: This top-level library was unused

File: js_src/_dart_helpers.js [hunk 4]
```diff
@@ -137,16 +137,16 @@ function _createReactDartComponentClass2(dartInteropStatics, componentStatics, j
...
-  jsConfig.skipMethods.forEach(method => {
-    if (ReactDartComponent2[method]) {
-      delete ReactDartComponent2[method];
-    } else {
-      delete ReactDartComponent2.prototype[method];
-    }
-  });
-
   if (jsConfig) {
+    // Delete methods that the user does not want to include (such as error boundary event).
```
Comment (greglittlefield-wf) [inline js_src/_dart_helpers.js:141]: I moved this so we could also verify `_createReactDartComponentClass2` is present when initializing

File: test/factory/js_factory_test.dart [hunk 5]
```diff
@@ -3,36 +3,56 @@
...
-      refTests<ReactComponent>(JsFoo, verifyRefValue: (ref) {
-        expect(isCompositeComponentWithTypeV2(ref, JsFoo), isTrue);
+      test('has a type corresponding to the backing JS class', () {
+        expect(JsFoo.type, equals(_JsFoo));
       });
     });
 
-    test('has a type corresponding to the backing JS class', () {
-      expect(JsFoo.type, equals(_JsFoo));
+    group('(function component)', () {
```
Comment (greglittlefield-wf) [inline test/factory/js_factory_test.dart:37]: This test case was mostly set up, just the tests for it were missing

File: test/factory/common_factory_tests.dart [hunk 6]
```diff
@@ -119,32 +119,38 @@ void commonFactoryTests(ReactComponentFactoryProxy factory, {bool isFunctionComp
...
+    final mountNode = Element.div();
+    react_dom.render(content, mountNode);
+    return mountNode.children.single;
+  }
+
   setUpAll(() {
     var called = false;
 
-    var renderedInstance = rtu.renderIntoDocument(factory({
+    final nodeWithClickHandler = renderAndGetRootNode(factory({
```
Comment (greglittlefield-wf) [inline test/factory/common_factory_tests.dart:131]: Updated to support function components, which don't have instances

**Comment by aaronlademann-wf** [comment]
+1 to https://github.com/cleandart/react-dart/pull/265/files/cf43b36..066c138 and https://github.com/cleandart/react-dart/pull/265/files/9714657..75d6611

**Pull request merged**